charmcraft/application/commands/lifecycle.py

```diff
@@ -0,0 +1,421 @@
+
+    @staticmethod
+    def _should_add_shell_args() -> bool:
+        return True
+
+
+class PullCommand(_LifecycleStepCommand):
+    """Command to pull parts."""
+
+    name = "pull"
+    help_msg = "Download or retrieve artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Download or retrieve artifacts defined for a part. If part names
+        are specified only those parts will be pulled, otherwise all parts
+        will be pulled.
+        """
+    )
+
+
+class BuildCommand(_LifecycleStepCommand):
+    """Command to build parts."""
+
+    name = "build"
+    help_msg = "Build artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Build artifacts defined for a part. If part names are specified only
+        those parts will be built, otherwise all parts will be built.
+        """
+    )
+
+
+class StageCommand(_LifecycleStepCommand):
+    """Command to stage parts."""
+
+    name = "stage"
+    help_msg = "Stage built artifacts into a common staging area"
+    overview = textwrap.dedent(
+        """
+        Stage built artifacts into a common staging area. If part names are
+        specified only those parts will be staged. The default is to stage
+        all parts.
+        """
+    )
+
+
+class PrimeCommand(_LifecycleStepCommand):
+    """Command to prime parts."""
+
+    name = "prime"
+    help_msg = "Prime artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Prepare the final payload to be packed, performing additional
+        processing and adding metadata files. If part names are specified only
+        those parts will be primed. The default is to prime all parts.
+        """
+    )
+
+    @override
+    def run(self, parsed_args: argparse.Namespace, step_name: str | None = None) -> None:
+        """Run the prime command."""
+        super().run(parsed_args, step_name=step_name)
+
+        self._services.package.write_metadata(self._services.lifecycle.prime_dir)
+
+
+class PackCommand(PrimeCommand):
+    """Command to pack the final artifact."""
+
+    name = "pack"
+    help_msg = "Build the charm or bundle"
+    overview = textwrap.dedent(
+        """
+        Build and pack a charm operator package or a bundle.
+
+        You can `juju deploy` the resulting `.charm` or bundle's `.zip`
+        file directly, or upload it to Charmhub with `charmcraft upload`.
+
+        For the charm you must be inside a charm directory with a valid
+        `metadata.yaml`, `requirements.txt` including the `ops` package
+        for the Python operator framework, and an operator entrypoint,
+        usually `src/charm.py`.  See `charmcraft init` to create a
+        template charm directory structure.
+
+        For the bundle you must already have a `bundle.yaml` (can be
+        generated by Juju) and a README.md file.
+        """
+    )
+
+    @override
+    def fill_parser(self, parser: argparse.ArgumentParser) -> None:
```

Comment:
These additional parser fills are to make `charmcraft pack` backwards compatible. The rest of the commands don't need this since charmcraft stable doesn't have any lifecycle commands except `pack` and `clean`